### PR TITLE
MERKLEBRANCHVERIFY

### PR DESCRIPTION
**Pull request by DesWurstes**
To speed up things before Feb 15: Bitcoin ABC rebase of Core's MAST PR (#12131, more details are there) by kallewoof. I tried to keep the attribution and stick to the format as much as I could. Tail call semantics PR? Soon.

The differences:
- MERKLEBRANCHVERIFY is U1<<17 while Maaku's is U1<<16. U1<<16 is Bitcoin Cash's SIGHASH_FORKID
- ~~I couldn't add this line: https://github.com/kallewoof/bitcoin/blob/461844dfbcf27118561293589c35b55436c49567/src/test/transaction_tests.cpp#L56~~

STATUS: Complete. Tests passing!

### CODE REVIEW

File: src/crypto/sha256.h [hunk 1]
```diff
@@ -20,8 +20,10 @@ class CSHA256 {
     static const size_t OUTPUT_SIZE = 32;
 
     CSHA256();
+    CSHA256(const unsigned char iv[OUTPUT_SIZE]);
     CSHA256 &Write(const uint8_t *data, size_t len);
     void Finalize(uint8_t hash[OUTPUT_SIZE]);
+    void Midstate(unsigned char hash[OUTPUT_SIZE], uint64_t* bytes, unsigned char* buffer);
```
Comment (gubatron) [inline src/crypto/sha256.h]: second parameter... is it `bytes` or is it `len`? one thing in the header, another in the implementation.

File: src/script/interpreter.cpp [hunk 2]
```diff
@@ -1183,6 +1185,142 @@ bool EvalScript(std::vector<valtype> &stack, const CScript &script,
...
+                            Unserialize(proofStream, branch.m_proof);
+                        } catch (const std::bad_alloc e) {
+                            throw; // Don't mask a transient out-of-memory exception
+                        } catch (...) {
+                            return set_error(serror, SCRIPT_ERR_BAD_DECODE_ARG3);
+                        }
+                        if (!proofStream.empty()) {
+                            // Extra bytes remaining after the MerkleProof
+                            // was deserialized, which could be a source
+                            // of witness malleability.
```
Comment (gubatron) [inline src/script/interpreter.cpp]: segwit logic leftovers? "witness malleability"

**Comment by schancel** [comment]
It would be great to get this formally proposed for the November protocol upgrade.  I'd really suggest joining the opcodes workgroup: https://github.com/bitcoincashorg/workgroups/blob/master/wg-opcodes/workgroup.md 

There are a number of other people there excited about future scripting additions.

**Comment by schancel** [comment]
@gubatron You should consider joining the development telegram :)

**Pull request closed without merging**